internal/controller/controllers/agent_controller.go

```diff
@@ -204,6 +207,113 @@ func (r *AgentReconciler) Reconcile(origCtx context.Context, req ctrl.Request) (
 }
 
+func findMatchingNode(agent *aiv1beta1.Agent, nodes []corev1.Node) *corev1.Node {
+	addresses := make([]string, 0)
+	for _, intf := range agent.Status.Inventory.Interfaces {
+		for _, ip := range append(intf.IPV4Addresses, intf.IPV6Addresses...) {
+			splitIp := strings.Split(ip, "/")
+			if len(splitIp) == 2 {
+				addresses = append(addresses, splitIp[0])
+			}
+		}
+	}
+	for i := range nodes {
+		n := &nodes[i]
+		if strings.EqualFold(n.Name, agent.Spec.Hostname) {
+			return n
+		}
+		for _, addr := range n.Status.Addresses {
+			if addr.Type == corev1.NodeInternalIP && funk.ContainsString(addresses, addr.Address) {
+				return n
+			}
+		}
+	}
+	return nil
+}
+
+func isNodeReady(node *corev1.Node) bool {
+	for _, c := range node.Status.Conditions {
+		if c.Status == corev1.ConditionTrue && c.Type == corev1.NodeReady {
+			return true
+		}
+	}
+	return false
+}
+
+func isNodeApproved(agent *aiv1beta1.Agent, nodes []corev1.Node) bool {
+	if agent.Status.DebugInfo.State == models.HostStatusAddedToExistingCluster &&
+		agent.Status.Progress.CurrentStage == models.HostStageRebooting {
+		node := findMatchingNode(agent, nodes)
+		if node != nil {
+			return isNodeReady(node)
+		}
+	}
+	return false
+}
+
+func (r *AgentReconciler) shouldApproveCSRs(agent *aiv1beta1.Agent, clients ClusterKubeClient) (bool, error) {
+
+	nodes, err := clients.ListNodes()
+	if err != nil {
+		return false, err
+	}
+	return !isNodeApproved(agent, nodes.Items), nil
+}
+func isCsrApproved(csr *certificatesv1.CertificateSigningRequest) bool {
+	for _, c := range csr.Status.Conditions {
+		if c.Type == certificatesv1.CertificateApproved {
+			return true
+		}
+	}
+	return false
+}
+
+func (r *AgentReconciler) ApproveAllCsrs(clients ClusterKubeClient) error {
+	csrList, err := clients.ListCsrs()
+	if err != nil {
+		r.Log.WithError(err).Error("Failed to get CSRs")
+		return err
+	}
+	for i := range csrList.Items {
+		csr := &csrList.Items[i]
+		if !isCsrApproved(csr) {
+			if err = clients.ApproveCsr(csr); err != nil {
+				r.Log.WithError(err).Error("Failed to approve CSR")
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+func (r *AgentReconciler) tryApproveDay2CSRs(ctx context.Context, agent *aiv1beta1.Agent) (alreadyApproved bool, requeue bool, err error) {
+	r.Log.Info("Approving CSRs")
+	namespacedName := types.NamespacedName{
+		Namespace: agent.Spec.ClusterDeploymentName.Namespace,
+		Name:      fmt.Sprintf(adminKubeConfigStringTemplate, agent.Spec.ClusterDeploymentName.Name),
+	}
+
+	secret, err := getSecret(ctx, r.Client, r.APIReader, namespacedName)
+	if err != nil {
+		return false, false, err
+	}
+	clients, err := r.ClusterKubeClientFactory.Create(secret)
+	if err != nil {
+		return false, false, err
+	}
+	shouldApprove, err := r.shouldApproveCSRs(agent, clients)
+	if err != nil {
+		return false, true, err
+	}
+
+	// Even if node is already ready, we try approving last time
+	err = r.ApproveAllCsrs(clients)
```

Comment:
Don't you need to approve only csrs related to a specific node?

Reply:
I will try to see if the CSRs can be associated with the specific agent.

Reply:
> I will try to see if the CSRs can be associated with the specific agent.

There should be a way to just approve the CSRs related to a specific host, in case we really need to approve these manually.

Reply:
one liner 
oc get csr csr-99n79 -o jsonpath='{ .spec.request }' | base64 -d |  openssl req  -in /dev/stdin -text  | grep  remote-worker1
        Subject: O=system:nodes, CN=system:node:remote-worker1